File: iosApp/iosApp/network/APIClient.swift
import Foundation
import sharedKit

final class APIClient {

    static let shared = APIClient()
    private init() {}

    private let session = URLSession.shared

    func request(
        url: URL,
        method: String = "GET",
        body: Data? = nil,
        requiresAuth: Bool = true
    ) async throws -> Data {

        var request = URLRequest(url: url)
        request.httpMethod = method
        request.httpBody = body
        request.setValue("application/json", forHTTPHeaderField: "Accept")

        if body != nil {
            request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        }

        if requiresAuth {
            guard let token = UserSessionProvider.shared.userSession.getAccessToken(),
                  !token.isEmpty else {
                throw APIError.unauthorized
            }
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }

        let (data, response) = try await session.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse else {
            throw APIError.invalidResponse
        }

        #if DEBUG
        print("STATUS:", httpResponse.statusCode)
        print(String(data: data, encoding: .utf8) ?? "No body")
        #endif

        switch httpResponse.statusCode {
        case 200...299:
            return data

        case 401:
            throw APIError.unauthorized

        case 403:
            throw APIError.forbidden

        case 500...599:
            throw APIError.serverError(httpResponse.statusCode)

        default:
            throw APIError.unknown
        }
    }
}

File: shared/src/commonMain/kotlin/com/example/shared/data/api/ApiClient.kt
package com.example.shared.data.api

import com.example.shared.data.keyValueStorage.UserSession
import com.example.shared.data.model.requests.RefreshTokenRequest
import com.example.shared.data.model.responses.auth.RefreshTokenResponse
import com.example.shared.log
import io.github.aakira.napier.Napier
import io.ktor.client.*
import io.ktor.client.call.*
import io.ktor.client.plugins.DefaultRequest
import io.ktor.client.plugins.HttpResponseValidator
import io.ktor.client.plugins.auth.Auth
import io.ktor.client.plugins.auth.providers.BearerTokens
import io.ktor.client.plugins.auth.providers.bearer
import io.ktor.client.request.*
import io.ktor.client.plugins.contentnegotiation.*
import io.ktor.client.plugins.logging.DEFAULT
import io.ktor.client.plugins.logging.LogLevel
import io.ktor.client.plugins.logging.Logger
import io.ktor.client.plugins.logging.Logging
import io.ktor.http.ContentType
import io.ktor.http.contentType
import io.ktor.serialization.kotlinx.json.*
import kotlinx.serialization.json.Json
import io.ktor.client.plugins.auth.providers.RefreshTokensParams


class ApiClient(
    private val userSession: UserSession
) {

    private val client = HttpClient {

        install(DefaultRequest) {
            header("Accept", "application/json")
        }

        install(Auth) {
            bearer {

                loadTokens {
                    log("init token") // 👈 ADD THIS
                    userSession.getAccessToken()?.let { token ->
                        log("DEBUG TOKEN FROM SESSION = $token") // 👈 ADD THIS
                        BearerTokens(
                            accessToken = token,
                            refreshToken = ""
                        )
                    }?:run {
                        log("No Bearer Token - log") // 👈 ADD THIS
                        null
                    }
                }

                refreshTokens {

                    val refreshToken = userSession.getUser()?.refreshToken
                        ?: return@refreshTokens null

                    try {
                        val response: RefreshTokenResponse =
                            client.post("${EcommerceApi.baseUrl}/auth/refresh") {
                                contentType(ContentType.Application.Json)
                                setBody(
                                    RefreshTokenRequest(refreshToken)
                                    //mapOf("refresh_token" to refreshToken)
                                )
                            }.body()

                        val user = userSession.getUser()
                        user?.token=response.data!!.access_token
                        userSession.saveUser(user!!)

                        BearerTokens(
                            accessToken = response.data!!.access_token,
                            refreshToken = refreshToken
                        )

                    } catch (e: Exception) {
                        print(e)
                        userSession.logout()
                        null
                    }
                }



                // 🔥 THIS IS REQUIRED
                sendWithoutRequest {
                    log("AUTH APPLIED TO: ${it.url}") // 👈 ADD THIS
                    Napier.d("AUTH APPLIED TO: ${it.url}") // 👈 ADD THIS
                    true
                }
            }
        }

        install(Logging) {
            logger = Logger.DEFAULT
            level = LogLevel.ALL
        }

        install(ContentNegotiation) {
            json(
                Json {
                    ignoreUnknownKeys = true
                    isLenient = true
                }
            )
        }

        HttpResponseValidator {
            validateResponse { response ->
                when (response.status.value) {
                    401 -> throw UnauthorizedException()
                    403 -> throw ForbiddenException()
                    in 500..599 -> throw ServerException()
                }
            }
        }
    }

    internal suspend inline fun <reified T> get(
        url: String,
        params: Map<String, String>? = null
    ): T =
        client.get(url) {
            contentType(ContentType.Application.Json)
            params?.forEach { (k, v) -> parameter(k, v) }
        }.body()

    internal suspend inline fun <reified T> post(
        url: String,
        body: Any?
    ): T =
        client.post(url) {
            contentType(ContentType.Application.Json)
            setBody(body)
        }.body()
}



class UnauthorizedException : Exception()
class ForbiddenException : Exception()
class ServerException : Exception()
